### ISSUE TITLE
Add Programmable NSE

### ISSUE THREAD

**Issue by brodieG** [Rdatatable/data.table#2432]
The idea would be to implement something like what [`oshka`](https://github.com/brodieG/oshka) does within `data.table`.  This would allow things like:
```
Jexp1 <- quote(max(conc))
Jexp2 <- quote(mean(conc))
Jexp <- quote(list(Jexp1, round(Jexp2)))

setDT(CO2)[, Jexp, by=Type]
##          Type max.conc. round.mean.conc..
## 1      Quebec      1000               435
## 2 Mississippi      1000               435
```
For more [details see the oshka vignette](https://htmlpreview.github.io/?https://github.com/brodieG/oshka/blob/master/inst/doc/nse-fun.html#an-ersatz-data.table) and the [twitter convo w/ @mattdowle on the topic](https://twitter.com/MattDowle/status/921168722173222913).

I need to look into whether there are any compatibility issues with the existing data table NSE.  One possible issue is that symbols that resolve to vectors in the parent frame have priority over variables inside the data.

For actual implementation we can just move code from `oshka` into `data.table`, or potentially re-write the `oshka` code in C although that may not be worth the trouble.

I have a couple of other projects ongoing now but once I wrap those up I'll circle back here with a more specific proposal.  In the meantime please let me know if you have thoughts/concerns/ideas.

**Comment by jangorecki** [comment]
Not exactly the same but related
https://github.com/Rdatatable/data.table/issues/1579

**Comment by brodieG** [comment]
I'd forgotten about this issue.  If I ever get interested in this again I'll resubmit something.

**Comment by jangorecki** [comment]
@brodieG I came across this issue, hoping it can be addressed by new proposed interface in #4304. Unfortunately what is blocking is that base R `substitute` cannot work on already quoted expressions, and we would need to fall back to some manual substitute over AST, as we do now to substitute LHS of `=` sign.
What works is to avoid nested substitution and always substitute when creating the next parent expression, like this:
```r
co2 = as.data.table(CO2)
Jexp1 = quote(max(conc))
Jexp2 = quote(mean(conc))
Jexp = substitute(list(Jexp1, round(Jexp2)), list(Jexp1=Jexp1, Jexp2=Jexp2))
co2[, j, by=Type, env=list(j=Jexp), verbose=TRUE]
#Argument 'by' after substitute: Type
#Argument 'j'  after substitute: list(max(conc), round(mean(conc)))
#...
#          Type    V1    V2
#        <fctr> <num> <num>
#1:      Quebec  1000   435
#2: Mississippi  1000   435
```
Your `oshka` package looks promising. Any idea if we could somehow incorporate this kind of recursive substitution reliably?

**Comment by brodieG** [comment]
I believe you can.  The code is reasonably small so in theory you could just copy it without having to add a dependency to the package.

I imagine you saw the vignette on making [a fake `data.table` with recursive substitution](https://htmlpreview.github.io/?https://github.com/brodieG/oshka/blob/master/inst/doc/nse-fun.html#an-ersatz-data.table)?

The biggest issue is that `oshka` hasn't seen much use, so who knows what un-anticipated corner cases are out there.  A couple of issues include that by recursively looking through the symbols, we touch every symbol involved.  This could trigger evaluation of promises or active bindings (Lionel brought this up in an unrelated discussion).  I think this is very much a corner case and can likely be mitigated with some measures if warranted (e.g. we can check for promises in C code, and we can check for active bindings in R code).

I can probably help a bit, if you give me a clearly defined description of what you're looking for I can try to put something together, although I'd want some sense of the probability that this actually gets used.
